google/cloud/bigquery/table.py

```diff
@@ -117,46 +118,136 @@ def _view_use_legacy_sql_getter(table):
+            self._properties, self._PROPERTY_TO_API_FIELD["project"]
+        )
+
+    @property
+    def dataset_id(self) -> str:
+        """ID of dataset containing the table."""
+        return _helpers._get_sub_prop(
+            self._properties, self._PROPERTY_TO_API_FIELD["dataset_id"]
+        )
+
+    @property
+    def table_id(self) -> str:
+        """The table ID."""
+        return _helpers._get_sub_prop(
+            self._properties, self._PROPERTY_TO_API_FIELD["table_id"]
+        )
+
+    @property
+    def path(self) -> str:
+        """URL path for the table's APIs."""
+        return (
+            f"/projects/{self.project}/datasets/{self.dataset_id}"
+            f"/tables/{self.table_id}"
+        )
+
+    def __eq__(self, other):
+        if isinstance(other, _TableBase):
+            return (
+                self.project == other.project
+                and self.dataset_id == other.dataset_id
+                and self.table_id == other.table_id
+            )
+        else:
+            return NotImplemented
+
+    def __hash__(self):
+        return hash((self.project, self.dataset_id, self.table_id))
+
+
+class TableReference(_TableBase):
     """TableReferences are pointers to tables.
 
     See
     https://cloud.google.com/bigquery/docs/reference/rest/v2/tables#tablereference
 
     Args:
-        dataset_ref (google.cloud.bigquery.dataset.DatasetReference):
+        dataset_ref:
             A pointer to the dataset
-        table_id (str): The ID of the table
+        table_id:
+            The ID of the table
+        project_id_alternative:
+            The alternative field that will be used when ESF is not able
+            to translate the received data to the project_id field.
+        dataset_id_alternative:
+            The alternative field that will be used when ESF is not able
+            to translate the received data to the project_id field.
+        table_id_alternative:
+            The alternative field that will be used when ESF is not able
+            to translate the received data to the project_id field.
     """
 
-    def __init__(self, dataset_ref, table_id):
-        self._project = dataset_ref.project
-        self._dataset_id = dataset_ref.dataset_id
-        self._table_id = table_id
+    _PROPERTY_TO_API_FIELD = {
+        "dataset_id": "datasetId",
+        "project": "projectId",
+        "table_id": "tableId",
+        "project_id_alternative": "projectIdAlternative",
+        "dataset_id_alternative": "datasetIdAlternative",
+        "table_id_alternative": "tableIdAlternative",
+    }
 
-    @property
-    def project(self):
-        """str: Project bound to the table"""
-        return self._project
+    def __init__(
+        self,
+        dataset_ref: "DatasetReference",
+        table_id: str,
+        project_id_alternative: Sequence[str] = None,
+        dataset_id_alternative: Sequence[str] = None,
+        table_id_alternative: Sequence[str] = None,
```

Comment:
These three "alternative" attributes are new, observed them in the generated code. Had to port them manually to the v3 refactored models.

Reply:
Weird. I'm investigating what their purpose are. From what I can tell they were added for weird API-translation layer reasons. Doubtful that they should ever have been exposed to the public.

Reply:
Found this comment "[When a BigQuery client] passes an array for some fields when it should be passing a single value, this currently works due to a bug".

Let's remove these fields. They are unnecessary and confusing.

Reply:
Done https://github.com/googleapis/python-bigquery/pull/996/commits/fc6f7fcf9873311c921ded4edda68d2c3d08aaed